File: 2019/17.jl
function part1(step::Int, cycles, target)
    state = [0]
    idx = 0
    for i in 1:cycles
        mod = (idx+step)%length(state)+1
        idx = mod==0 ? length(state) : mod
        state = insert!(state, idx, i)
    end

    idx = findfirst(x->x==target, state)
    println("🎄 $(state[idx-3:idx+4])")
end

function part2(step::Int, cycles)
    # since 0 is always last it is enough to check who is first
    # in the sequence
    # also the first element will change every time 
    # (step + idx)%current_length == 0

    idx = 0
    val = 0
    for i in 1:cycles
        idx = (idx+step)%i + 1
        if idx == 1
            val = i
        end
    end
    println("🎁🎄 $(val)")
end

part1(382, 2017, 2017)
part2(382, 50000000)

File: 2017/17.py

step = 382
state = [0]
idx = 0
for i in range(1, 2018):
    idx = (idx+step)%len(state) + 1
    state.insert(idx, i)

idx = state.index(2017)
print ("🎄Part 1: {}".format((state[idx-3:idx+4])))

idx = 0
for i in range(1, 50000000+1):
    idx = (idx+step)%i + 1
    if idx == 1:
        val = i
    
print ("🎁🎄Part 2: {}".format(val))
